### ISSUE TITLE
NameSpaceRuleChecker: Questions and possible bug with DELETE

### ISSUE THREAD

**Issue by rennhard** [microsoft/restler-fuzzer#104]
Dear developers

I'm aware that this checker is currently off by default, but nevertheless I played around with it and have a few questions about it. It's basically three questions in one issue, but they are related to each other so I'm including them in one issue.

To test, I'm using the following REST API:

**POST /rest/admin/products/{username}**
=> This creates a product provided by user {username}. This can only be done by a user with a valid JSON Web Token (JWT) and only if the username in the token corresponds to {username}. It returns, among other data, the productID of the newly created product.

**GET  /rest/admin/products/{username}/{productID}**
=> Allows to get a product. Can only be done with valid JWT that contains the corresponding username. So just like the POST request above, this is part of the "administrative area" of the API (anything below /rest/admin) to maintain products.

**DELETE  /rest/admin/products/{username}/{productID}**
=> Allows to delete a product. Can only be done with valid JWT that contains the corresponding username. Also part of the "administrative area".

**GET  /rest/products/{productID}**
=> Allows any user (no authentication required) to get a product. So in contrast to the requests above, this is not part of the "administrative area" of the API used for maintaining product, but allows anonymous users to view product details.

There's a little python program that returns valid JWTs of two users, luke and snoopy. This is used in parameter --token_refresh_command.

```
#!/usr/bin/env python
import sys
sys.stdout.write("{'luke':{}, 'snoopy':{}}\n")
sys.stdout.write("Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJNYXJrZXRwbGFjZSIsInN1YiI6Imx1a2UiLCJleHAiOjE5MDczNTM5NzJ9.jdoYfz_diOQa...\n")
sys.stdout.write("Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJNYXJrZXRwbGFjZSIsInN1YiI6InNub29weSIsImV4cCI6MTkwNzM1NDAxMX0.XiOtGu...\n")
```

I've added two vulnerabilities:
- With a valid JWT, one can GET the products of another user, e.g., snoopy can successfully execute GET  /rest/admin/products/luke/25 to get a product that belongs to luke.
- With a valid JWT, one can DELETE the products of another user, e.g., snoopy can successfully execute DELETE  /rest/admin/products/luke/25 to delete a product that belongs to luke.

The results of fuzz-lean are as follows:

- GET /rest/admin/products/luke/25 as user snoopy is correctly recognised as a vulnerability.
- GET /rest/products/25 is wrongly flagged as a vulnerability. Intuitively, I wouldn't have expected this as the POST request created the product below /rest/admin/products/luke. But apparently, a vulnerability is flagged if the corresponding productID is used in any other request, no matter what the full pat is. Is this intentional?
-  DELETE /rest/admin/products/luke/25 as user snoopy is not recognised as a vulnerability. Apparently, when testing this, restler first does a POST and a corresponding DELETE (so the product is removed), and then it does another DELETE (I assume with the second user). This results in a 4xx error, as the product is longer there. This seems to be a wrong verification approach and a possible bug. The correct approach would be to do POST with the JWT of user1 and then DELETE with the JWT of user2.

And a final question: Is there a way to link the username in a token (that is returned by the python script above in my case) with the username that is used in the path of the POST request? Right now, I have hardcoded this in Grammar.py which works (so it always uses username luke), but this is not really an elegant solution.

So to summarise: three questions/issues:
- Is allowed access to GET /rest/products/25 intentional?
- Is the missed vulnerability with DELETE /rest/admin/products/luke/25 due to a bug?
- Is there a good way to link the username in a JWT to parameters used in requests?

For reference, I've attached Grammar.py, the network log and the two reported vulnerabilities.

Thanks a lot!
Marc
[grammar.py.zip](https://github.com/microsoft/restler-fuzzer/files/5657478/grammar.py.zip)
[network.testing.140572845729536.1.txt](https://github.com/microsoft/restler-fuzzer/files/5657479/network.testing.140572845729536.1.txt)
[NameSpaceRuleChecker_20x_1.txt](https://github.com/microsoft/restler-fuzzer/files/5657480/NameSpaceRuleChecker_20x_1.txt)
[NameSpaceRuleChecker_20x_2.txt](https://github.com/microsoft/restler-fuzzer/files/5657481/NameSpaceRuleChecker_20x_2.txt)

**Comment by PatGod** [comment]
@rennhard Hi Marc, thanks for your detailed feedback and questions. Here are some answers.

> Is allowed access to GET /rest/products/25 intentional?

Yes, by design, the NameSpaceRule checker intentionally flags this access by user-2 of a dynamically-created resource created by user-1. This is to err on the safe side: the checker reports all such (potentially dangerous) valid accesses, at the risk of reporting false alarms in some contexts. 

> Is the missed vulnerability with DELETE /rest/admin/products/luke/25 due to a bug?

If you set `"mode": exhaustive` in this checker settings (in the settings file - see https://github.com/microsoft/restler-fuzzer/blob/main/docs/user-guide/Checkers.md for how to do this and the exact syntax), you should then see the sequence "POST by user-1;DELETE by user-2" that you expected. (If not, please let us know.)

> Is there a good way to link the username in a JWT to parameters used in requests?

No, as what you call "username" in your example is implicit/nameless with RESTler and this checker - the first JWT can dynamically create some resources, while the second JWT tries to access those and reports an alarm whenever it is successful. (The 'names' associated with JWT1 and JWT2, if any, are irrelevant.) 

Please let us know if this does not address adequately your questions. Thanks!

**Comment by rennhard** [comment]
Hi Patrice

Thanks a lot for the feedback. Setting `"mode":"exhaustive"` does not help to detect the DELETE issue. I'm using the following setting file:

```
{
    "checkers": {
        "invaliddynamicobject": {
            "no_defaults":true
        },
        "namespacerule" : {
            "mode":"exhaustive"
        }
    }
}
```

Note that when adding a bug to the REST API so that when calling DELETE the resource is not deleted, then the issue (sequence POST user1, DELETE user2) is correctly detected (as DELETE now returns 200 and not a 4xx status code because the resource is still existing), even without exhaustive mode. The output of the checker is attached. So the problem may indeed be that the NameSpaceRuleChecker uses a wrong approach when checking the sequence POST - DELETE in the sense that it tries to re-DELETE a resource that is no longer there?

Thanks,
Marc

[NameSpaceRuleChecker_20x_1.txt](https://github.com/microsoft/restler-fuzzer/files/5683085/NameSpaceRuleChecker_20x_1.txt)

**Comment by PatGod** [comment]
Can you please attach the full network.testing.xxx logs?

**Comment by rennhard** [comment]
I'm adding two network logs:

- network.testing.140006863460096.1.txt: This is based on the REST API that does not actually delete the resource when DELETE is called (but it still returns status code 204). Here, the NameSpaceRule Checker finds the vulnerability with the sequence POST-DELETE. And there's also a vulnerability reported by the UseAfterFree checker because the resource is not deleted.

- network.testing.139987614951168.1.txt: Here, the REST API correctly deletes the resource when DELETE is called. The NameSpaceRule Checker does not find the vulnerability (although it should).

[network.testing.140006863460096.1.txt](https://github.com/microsoft/restler-fuzzer/files/5684704/network.testing.140006863460096.1.txt)
[network.testing.139987614951168.1.txt](https://github.com/microsoft/restler-fuzzer/files/5684705/network.testing.139987614951168.1.txt)

**Comment by PatGod** [comment]
Thanks for these network logs. But can you please confirm that these logs were obtained with
```
"namespacerule" : {
            "mode":"exhaustive"
        }
```
? (They do not look like it...) Thanks again!

**Comment by rennhard** [comment]
I'm pretty sure I'm using exhaustive mode. I'm using the following file `user-settings.json` on the command line with `--settings user-settings.json`:

```
{
    "checkers": {
        "invaliddynamicobject": {
            "no_defaults":true
        },
        "namespacerule" : {
            "mode":"exhaustive"
        }
    }
}
```

The reason why I also added the rule for invaliddynamicobject was to make sure the settings file is working (as I get such a vulnerability without the settings file and it went away when using the settings file), as the output files of fuzz-lean don't contain any information about what checker settings were actually used.

I also just checked with fuzz mode (instead of fuzz-lean), but this has no effect on vulnerabilities that are reported (or not) with the DELETE request.

**Comment by rifiles** [comment]
Thanks Marc.  I've created a new issue to stop the checker from replaying the final request in the sequence prior to sending the attack request. This should hopefully relieve the issue surrounding false negatives when attacking with DELETE requests. I'll look more into what's going on with exhaustive mode as well.

**Comment by PatGod** [comment]
Thanks again Marc. 

To be clear, there are 2 distinct action items:
- the Namespacerule checker should test the 'POST;DELETE' sequence even in non-exhaustive mode - new issue #112 has been filed to address this. 
- the network logs above do not show all the consumer requests being tried in exhaustive mode (as we had expected) - this should be investigated as well (either the exhaustive mode was not turned on properly, or there is a glitch somewhere).

**Comment by rennhard** [comment]
Thanks a lot for taking care of this.

I'm closing this issue.

**Comment by PatGod** [comment]
Just to conclude: 
- the first point above was fixed with issue #112
- the second point above: after further investigation, my understanding of the exhaustive mode was wrong (my mistake) and there was no glitch here.

Thanks again for your feedback.
